### ISSUE TITLE
[MIUI-Android]The widget’s Hyperlink cannot open the URL

### ISSUE THREAD

**Issue by jiqi136** [fyne-io/fyne#1514]
### Describe the bug:

The official example generated [Android] APK can not open the URL link
https://github.com/fyne-io/fyne/tree/master/cmd/fyne_demo

I tried each of the following methods, but URL opening failed

=======

`
package main

import (
	"net/url"  
	"log"
	"os"
	"fyne.io/fyne/app" 
	"fyne.io/fyne/widget"	
)

func main() {
	var giturl= "http://github.com"
	
	log.Println("----mo一一Command ， -----》")	
	mo一一Command(giturl)
	
	var Newapp = app.New()
	var myWindow = Newapp.NewWindow("url") 

	var git, _ = url.Parse(giturl)
	_ =Newapp.OpenURL(git)
	
	var widgeturl= widget.NewHyperlink("giturl",git)
	
	myWindow.SetContent(widgeturl) 
	myWindow.Show()
	Newapp.Run() 
}

func mo一一Command(giturl string){
	var cmd1 = exec.Command("am", "start", "-a", "android.intent.action.VIEW","--user", "0","-d", giturl)
	var cmd2 = exec.Command("/system/bin/am", "start", "-a", "android.intent.action.VIEW","-d", giturl)
	var cmd3 = exec.Command("am", "start", "-a", "android.intent.action.VIEW","-d", giturl)
	var cmd4 = exec.Command("am", "start", "-a", "android.intent.action.VIEW","-c","android.intent.category.BROWSABLE","-d", giturl)
	var cmd5 = exec.Command("am", "start", "-a", "android.intent.action.VIEW","-c","android.intent.category.DEFAULT","-d", giturl)
	var cmd6 = exec.Command("am", "start", "-a", "android.intent.action.VIEW","-c","android.intent.category.BROWSABLE","-d", giturl,"org.mozilla.firefox")
	var cmd7 = exec.Command("am", "start", "-a", "android.intent.action.VIEW","-c","android.intent.category.DEFAULT","-d", giturl,"org.mozilla.firefox")
	var cmd8 = exec.Command("am", "start", "-a", "android.intent.action.VIEW","-d", giturl,"com.android.browser")
	
	for _, cmd := range []*exec.Cmd{cmd1,cmd2,cmd3,cmd4,cmd5,cmd6,cmd7,cmd8} { 
		cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
		var err=cmd.Run()
		if err != nil {
			//No error report
			log.Println("err",err) 
		}
	}
	return 
}`

======================
AndroidManifest.xml 
File added 100 permissions

`<uses-permission android:name="android.permission.INTERNET" />  `

`<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>`

....
....
....

### Device (please complete the following information):
 - OS:Android
 - Version: 4.4
 - Go version:1.13
 - Fyne version: 1.14

**Comment by andydotxyz** [comment]
Can you please provide the errors that were encountered?
You should not need any permissions to access that function, but perhaps your version of android does not ship "am".
Can you also please provide more information about your device? There are many differences possible inside the "Android 4.4" umbrella.

**Comment by jiqi136** [comment]
@Andydotxyz
The tested Redmi 2 mobile phone is Android version 4.4. Because the mobile phone driver cannot be installed on the computer, adb cannot connect to the mobile phone and cannot get the record information of the error reason.

The "am" function can be used on the mobile phone, and the "am" function can be used to successfully create folders and delete files.
But the "am" command is invalid when opening files, devices or other applications. It seems to be limited to the set permissions.

To traverse, use the previous source code "am" command, which is "err== nil".

`var err=cmd.Run()
if err != nil {
No error report
}`

**Comment by andydotxyz** [comment]
I cannot replicate this issue, so I am asking for your help.
What is the output of your println? `log.Println("err",err)`

**Comment by jiqi136** [comment]
@Andydotxyz

"Am" runs without error, "err == nil".
But the "am" command is invalid when opening files, devices or other applications.

**Comment by andydotxyz** [comment]
There must be some indication of it not working... if error is nil then perhaps there is some info on StdErr or StdOut?

**Comment by stuartmscott** [comment]
> Because the mobile phone driver cannot be installed on the computer, adb cannot connect to the mobile phone and cannot get the record information of the error reason.

This issue should be fixed first - we cannot help diagnose the error without knowing why it is breaking. Can you try with android emulator? 

> ```go
> var git, _ = url.Parse(giturl)
> _ =Newapp.OpenURL(git)
> ```

You should not ignore errors like this.

**Comment by jiqi136** [comment]
@Andydotxyz   @stuartmscott
After some hard work, I finally ran ABD and saved the record file.

**1.The record corresponding to "am start"**
var cmd = exec.Command("am", "start", "-a", "android.intent.action.VIEW","-d", giturl)

`I/Fyne    ( 9402): 26 -------mo一一Command------->
D/VoldCmdListener(  151): CommandListener runCommand mkdir argv[2]:/storage/sdcard1/Android/data/org.golang.todo.a5/files/
W/Vold    (  151): Returning OperationFailed - no handler for errno 30
D/AndroidRuntime( 9423): 
D/AndroidRuntime( 9423): >>>>>> AndroidRuntime START com.android.internal.os.RuntimeInit <<<<<<
D/AndroidRuntime( 9423): CheckJNI is OFF
E/dalvikvm( 9423): ERROR: must specify non-'.' bootclasspath
W/dalvikvm( 9423): CreateJavaVM failed: dvmClassStartup failed
E/AndroidRuntime( 9423): JNI_CreateJavaVM failed
I/Fyne    ( 9402): 29 mo一一Command end <----`

**2.The record corresponding to "Newapp.OpenUR"**

`I/Fyne    ( 9402): 37 -------Newapp.OpenUR------->
D/AndroidRuntime( 9424): 
D/AndroidRuntime( 9424): >>>>>> AndroidRuntime START com.android.internal.os.RuntimeInit <<<<<<
D/AndroidRuntime( 9424): CheckJNI is OFF
E/dalvikvm( 9424): ERROR: must specify non-'.' bootclasspath
W/dalvikvm( 9424): CreateJavaVM failed: dvmClassStartup failed
E/AndroidRuntime( 9424): JNI_CreateJavaVM failed
I/Fyne    ( 9402): 37 Newapp.OpenUR  end <---  
`

**3."widget.NewHyperlink"对应的记录**
查看上传的附件

[test.log](https://github.com/fyne-io/fyne/files/5512776/test.log)

**Comment by stuartmscott** [comment]
This is progress, now we know the actual issue is:

> `E/dalvikvm( 9424): ERROR: must specify non-'.' bootclasspath`

Is this using a physical device or the android emulator?

**Comment by jiqi136** [comment]
Real Android phone operation and finger click operation

**Comment by stuartmscott** [comment]
Try with the emulator so we know whether it is an issue with your device or not

**Comment by jiqi136** [comment]
@stuartmscott

Tri markoj de veraj lastaj telefonoj, povas lanĉi kaj malfermi la retejon

VMware virtuala maŝinoinstalo Android-x86-5.1-rc1.iso,
Kvankam la operacio falis,
Sed antaŭ la krazo, ĝi ankaŭ finis malfermi la URL

**Comment by andydotxyz** [comment]
What is special about this device I wonder? What is the manufacturer/model and are there any operator-customisations loaded?

**Comment by jiqi136** [comment]
@Andydotxyz

Mobile phone name: Hongmi mobile phone
Mobile phone model: HM2a
Android version: 4.4.4KTU84Q
MIUI version: MIUI9.2

**Comment by andydotxyz** [comment]
Ah, so the MIUI is not really Android I think. 
The classpath configuration may be incompatible with certain standard actions - and it looks like "am" is one of those things that does not work.
I guess we need to see if there is an Android API that MIUI still supports that we can use instead

**Comment by donatj** [comment]
For what it's worth I am having the same problem on my Google Pixel 4

**Comment by donatj** [comment]
```
url, _ := url.Parse("https://www.google.com")
err := fyne.CurrentApp().OpenURL(url)
fmt.Println(err);
```

gets me

```
exit status 255
```

**Comment by andydotxyz** [comment]
That is very helpful thank you @donatj.

Is it possible for you find the appropriate entry in your android log?
Before the 255 exit there should be output printed. This content will be in the Android log (`adb logcat`) with a `Fyne` or `Go` prefix...

**Comment by stuartmscott** [comment]
I just hit this too on Pixel 2 w/ Android 11;

```
02-19 08:00:42.814 15898 15935 I Fyne    : Starting: Intent { act=android.intent.action.VIEW dat=https://space-nyc.aletheiaware.com/... }
--------- beginning of system
02-19 08:00:42.815  1427  2064 W ActivityTaskManager: Permission Denial: package=com.android.shell does not belong to uid=10298
02-19 08:00:42.815 15898 15931 E Fyne    : 
02-19 08:00:42.815 15898 15931 E Fyne    : Exception occurred while executing 'start':
02-19 08:00:42.815 15898 15931 E Fyne    : java.lang.SecurityException: Permission Denial: package=com.android.shell does not belong to uid=10298
02-19 08:00:42.815 15898 15931 E Fyne    : 	at com.android.server.wm.ActivityTaskManagerService.assertPackageMatchesCallingUid(ActivityTaskManagerService.java:2598)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1081)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1072)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at com.android.server.am.ActivityManagerService.startActivityAsUserWithFeature(ActivityManagerService.java:3678)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at com.android.server.am.ActivityManagerShellCommand.runStartActivity(ActivityManagerShellCommand.java:544)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at com.android.server.am.ActivityManagerShellCommand.onCommand(ActivityManagerShellCommand.java:186)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at android.os.BasicShellCommandHandler.exec(BasicShellCommandHandler.java:98)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at android.os.ShellCommand.exec(ShellCommand.java:44)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at com.android.server.am.ActivityManagerService.onShellCommand(ActivityManagerService.java:10522)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at android.os.Binder.shellCommand(Binder.java:929)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at android.os.Binder.onTransact(Binder.java:813)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:5027)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at com.android.server.am.ActivityManagerService.onTransact(ActivityManagerService.java:2883)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at android.os.Binder.execTransactInternal(Binder.java:1159)
02-19 08:00:42.815 15898 15931 E Fyne    : 	at android.os.Binder.execTransact(Binder.java:1123)
02-19 08:00:42.815 22360 22362 I cmd     : oneway function results will be dropped but finished with status OK and parcel size 4
02-19 08:00:42.817 15898 15930 I Fyne    : 2021/02/19 08:00:42 fyne.go:491: Error: exit status 255
```

**Comment by AlbinoGeek** [comment]
> I just hit this too on Pixel 2 w/ Android 11

Are we doing this like we should be?

> You need to set android:exported="true" in your `AndroidManifest.xml` file where you declare this Activity.

**Comment by donatj** [comment]
Finally figured out how to use logcat (sorry, I'm not primarily an Android dev)

```
02-19 11:13:59.703 13176 13237 I Fyne    : Starting: Intent { act=android.intent.action.VIEW dat=https://github.com/... }
02-19 11:13:59.703 13176 13237 E Fyne    : 
02-19 11:13:59.703 13176 13237 E Fyne    : Exception occurred while executing 'start':
02-19 11:13:59.703 13176 13237 E Fyne    : java.lang.SecurityException: Permission Denial: package=com.android.shell does not belong to uid=10373
02-19 11:13:59.703 13176 13237 E Fyne    : 	at com.android.server.wm.ActivityTaskManagerService.assertPackageMatchesCallingUid(ActivityTaskManagerService.java:2594)
02-19 11:13:59.703 13176 13237 E Fyne    : 	at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1077)
02-19 11:13:59.703 13176 13237 E Fyne    : 	at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1068)
02-19 11:13:59.703 13176 13237 E Fyne    : 	at com.android.server.am.ActivityManagerService.startActivityAsUserWithFeature(ActivityManagerService.java:3662)
02-19 11:13:59.703 13176 13237 E Fyne    : 	at com.android.server.am.ActivityManagerShellCommand.runStartActivity(ActivityManagerShellCommand.java:544)
02-19 11:13:59.703 13176 13237 E Fyne    : 	at com.android.server.am.ActivityManagerShellCommand.onCommand(ActivityManagerShellCommand.java:186)
02-19 11:13:59.703 13176 13237 E Fyne    : 	at android.os.BasicShellCommandHandler.exec(BasicShellCommandHandler.java:98)
02-19 11:13:59.703 13176 13237 E Fyne    : 	at android.os.ShellCommand.exec(ShellCommand.java:44)
02-19 11:13:59.703 13176 13237 E Fyne    : 	at com.android.server.am.ActivityManagerService.onShellCommand(ActivityManagerService.java:10505)
02-19 11:13:59.703 13176 13237 E Fyne    : 	at android.os.Binder.shellCommand(Binder.java:929)
02-19 11:13:59.704 13176 13237 E Fyne    : 	at android.os.Binder.onTransact(Binder.java:813)
02-19 11:13:59.704 13176 13237 E Fyne    : 	at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:5053)
02-19 11:13:59.704 13176 13237 E Fyne    : 	at com.android.server.am.ActivityManagerService.onTransact(ActivityManagerService.java:2867)
02-19 11:13:59.704 13176 13237 E Fyne    : 	at android.os.Binder.execTransactInternal(Binder.java:1159)
02-19 11:13:59.704 13176 13237 E Fyne    : 	at android.os.Binder.execTransact(Binder.java:1123)
02-19 11:13:59.704 14282 14284 I cmd     : oneway function results will be dropped but finished with status OK and parcel size 4
02-19 11:13:59.707 13176 13266 I Fyne    : Fyne error:  Failed to open url
02-19 11:13:59.707 13176 13266 I Fyne    :   Cause: exit status 255
02-19 11:13:59.707 13176 13266 I Fyne    :   At: /Users/jdonat/gocode/pkg/mod/fyne.io/fyne/v2@v2.0.0/widget/hyperlink.go:115
```

**Comment by andydotxyz** [comment]
Would it be possible to run a simple app on one of these devices to launch it using the activity API like this:
https://stackoverflow.com/questions/2201917/how-can-i-open-a-url-in-androids-web-browser-from-my-application

That might help track down if it is solvable by removing dependency on the helper tool we invoke.

**Comment by andydotxyz** [comment]
Resolved on `develop` branch for v2.1, apologies for the delay!
